ec2 client credential init fix for AWS IRSA

Greyeye opened the pull request:
### What does this PR do?
This PR modifies the initialisation of the AWS EC2 client to use the AWS configuration package. This change enables support for built-in credentials such as AWS IAM Roles for Service Accounts (IRSA). The current use of ec2.New() does not work with IRSA, and although not tested, a similar issue likely occurs with EKS Pod Identity as well.

When a pod is assigned an IRSA with the correct permissions to perform DescribeTags, it encounters an error and logs the following message:

```
(pkg/util/ec2/ec2_tags.go:104 in fetchEc2TagsFromAPI) | unable to get tags using default credentials (falling back to instance role): operation error EC2: DescribeTags, https response error StatusCode: 400, RequestID: 1234-1234-1234, api error MissingParameter: The request must contain the parameter AWSAccessKeyId
```
fixes [issue 29916](https://github.com/DataDog/datadog-agent/issues/29916)

### Motivation
We are currently running the Datadog Agent pod in AWS EKS and are aiming to clean up the instance role by eliminating unnecessary permissions attached to it. Datadog is one of many services that currently use the instance role for the ec2:DescribeTags permission, and we would like to switch to using the IRSA-assigned role.

After extensive debugging and tracing, we discovered that when the pod is assigned an IRSA token but the instance is running IMDSv2, it fails to use IRSA and always falls back to using the instance role.

### Describe how to test/QA your changes
1. create a new EKS node and instance role that does not allow ec2:describeTags.
2. create a new AWS IAM Rolewith ec2:describeTags, and attach this to datadog-agent using IRSA
3. deploy datadog-agent with config.yaml (or using configMap) like this 
collect_ec2_tags: true
collect_ec2_tags_use_imds: true
ec2_prefer_imdsv2: true
4. start the datadog-agent and verify the role gets used, and tags are collected using the IRSA role.

### Possible Drawbacks / Trade-offs
none

### Additional Notes
none

Greyeye committed:
ec2 client credential init fix

current use of ec2.New() does not work with IRSA. (not tested but most like same problem with EKS Pod Identity as well)

```
(pkg/util/ec2/ec2_tags.go:104 in fetchEc2TagsFromAPI) | unable to get tags using default credentials (falling back to instance role): operation error EC2: DescribeTags, https response error StatusCode: 400, RequestID: 1234-1234-1234, api error MissingParameter: The request must contain the parameter AWSAccessKeyId
```

This change is to change to use aws config package to help initialise aws ec2 client with the built-in credentials such as AWS IRSA tokens.

Greyeye committed:
add release notes

louis-cqrl commented on pkg/util/ec2/ec2_tags.go:
```diff
@@ -111,14 +112,25 @@ func fetchEc2TagsFromAPI(ctx context.Context) ([]string, error) {
+		log.Warnf("unable to get aws configurations: %s", err)
+		return nil
+	}
+	return ec2.NewFromConfig(cfg)
```
❓ maybe add a log such as: `log.Info("AWS configurations successfully fetched")`
What do you think ?

Greyeye replied:
@louis-cqrl I believe that will populate unnecessary log output. How about log.Debug() instead.

louis-cqrl commented on pkg/util/ec2/ec2_tags.go:
```diff
@@ -111,14 +112,25 @@ func fetchEc2TagsFromAPI(ctx context.Context) ([]string, error) {
+	return ec2.NewFromConfig(cfg)
+}
+
+func getTagsWithCreds(ctx context.Context, instanceIdentity *EC2Identity, connection ec2ClientInterface) ([]string, error) {
```
❓ if `ec2Cconnection` fails then `connection.DescribeTags` will unreference a nil pointer, isn't an issue ?

Greyeye replied:
That's a good find, I've added the check and updated the test

louis-cqrl commented on pkg/util/ec2/ec2_tags.go:
```diff
@@ -111,14 +112,25 @@ func fetchEc2TagsFromAPI(ctx context.Context) ([]string, error) {
+}
+
+// ec2Cconnection creates an ec2 client with the given region and credentials
+func ec2Cconnection(ctx context.Context, region string, awsCreds aws.CredentialsProvider) ec2ClientInterface {
```
❓ Why the function is name `ec2Cconnection` and not `ec2Connection` ?

Greyeye replied:
this has been fixed

Greyeye committed:
review fixes

1. rename function name ec2Cconnection to ec2Connection
2. add nil check for connection
3. updated the unit test to cover nil test
4. add debug log print for aws config setup

drichards-87 commented on releasenotes/notes/ec2-cred-fix-9326728a245381e7.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+enhancements:
+  - |
+    Modifies the initialisation of the AWS EC2 client to use the AWS configuration package.
+    This change enables support for built-in credentials such as AWS IAM Roles for Service Accounts (IRSA).
```
```suggestion
    Modifies the initialization of the Amazon EC2 client to use the AWS configuration package.
    This change enables support for built-in credentials, such as AWS IAM Roles for Service Accounts (IRSA).
```

drichards-87 reviewed (approved):
Left some minor feedback from Docs and approved the PR.

Greyeye committed:
Update releasenotes/notes/ec2-cred-fix-9326728a245381e7.yaml

Co-authored-by: DeForest Richards <56796055+drichards-87@users.noreply.github.com>

mpescetto-omada commented:
Is there a time frame for this to be picked back up?

louis-cqrl commented:
Hello can you pull main and fix the conflicts on your branch in order to proceed to the next steps on our side ?

s-alad commented:
Should be addressed in https://github.com/DataDog/datadog-agent/pull/45530
